test/unit/input_template_renderer_test.rb

```diff
@@ -95,6 +95,147 @@
+            renderer.error_message.must_include 'Recursive rendering of templates detected'
+          end
+
+          it 'handles circular references in inputs' do
+            assert_raises(ActiveRecord::RecordInvalid) do
+              input_set = FactoryGirl.build(:foreign_input_set, :target_template => template_with_inputs, :include_all => false,
+                                                                :include => 'package, debug', :exclude => 'action,debug')
+              recursive_template_with_inputs.foreign_input_sets << input_set
+              recursive_template_with_inputs.save!
+            end
+          end
+        end
+      end
+    end
+
+    context 'renderer for template with input set and render_template' do
+      let(:command_template) do
+        FactoryGirl.build(:job_template, :name => 'command action', :template => '<%= input("command") -%>').tap do |template|
+          template.template_inputs << FactoryGirl.build(:template_input, :name => 'command', :input_type => 'user')
+          template.template_inputs << FactoryGirl.build(:template_input, :name => 'debug', :input_type => 'user')
+        end
+      end
+
+      let(:package_template) do
+        FactoryGirl.build(:job_template, :name => 'package action', :template => <<TEMPLATE)  do |template|
+<%= render_template("command action", "command" => "yum -y \#{ input("action") } \#{ input('package') }") -%>
+TEMPLATE
+          template.template_inputs << FactoryGirl.build(:template_input, :name => 'package', :input_type => 'user')
+          template.template_inputs << FactoryGirl.build(:template_input, :name => 'action', :input_type => 'user')
+          template.foreign_input_sets << FactoryGirl.build(:foreign_input_set, :target_template => command_template, :include_all => true, :exclude => 'command')
+        end
+      end
+
+      let(:template) do
+        FactoryGirl.create(:job_template,
+                           :template => '<%= render_template("package action", "action" => "install", :with_foreign_input_set => true) %>').tap do |template|
+          template.foreign_input_sets << FactoryGirl.build(:foreign_input_set, :target_template => package_template, :include_all => true, :exclude => 'action')
+        end
+      end
+
+      let(:job_invocation) { FactoryGirl.create(:job_invocation) }
+      let(:template_invocation) { FactoryGirl.build(:template_invocation, :template => template) }
+      let(:renderer) { InputTemplateRenderer.new(template) }
+      let(:result) { renderer.render }
+
+      before do
+        User.current = users :admin
+        command_template.save!
+        package_template.save!
+        job_invocation.template_invocations << template_invocation
+      end
+
+      describe 'foreign input set' do
+        describe 'with include_all' do
+          let(:template) do
+            FactoryGirl.create(:job_template, :template => '<%= render_template("package action", "action" => "install) %>').tap do |template|
```

Comment:
would fail on missing `"` after `install` but it seems you only test inputs that were found :-)